src/click/core.py

```diff
@@ -2618,6 +2618,12 @@ def __init__(
             # default.
             self.type = types.convert_type(None, flag_value)
 
+        if is_flag and isinstance(self.type, types.BoolParamType):
+            # If the type is a boolean, we need to set the flag_value
+            # to True or False depending on the default.
+            if flag_value is None:
+                flag_value = not self.default
+
```

Comment:
this code was nested in https://github.com/pallets/click/pull/2829 causing the current error.

Click was executing it in case any of the previous conditions were not met but after nesting it, it was no longer reachable.

To not break the previous change I added `isinstance(self.type, types.BoolParamType)` so we preserve the introduced behaviour in `2829`

Reply:
If you do not use any type for click flags, click will automatically detect the flag and everything will work as it was before the breaking change working